### PR TITLE
Handle exception while opening connection from constructor

### PR DESCRIPTION
**Pull request by arunmk-ms**
Fixes #.
[Issue 934](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/934)

## Changes
Add try/catch to the socket connection in constructor so that the application wont fail fast when the UDS is not available at startup.

### CODE REVIEW

**Review by reyang — APPROVED** [review]
LGTM with a changelog entry and clearing the CLA.

File: src/OpenTelemetry.Exporter.Geneva/CHANGELOG.md
Comment (reyang) [inline src/OpenTelemetry.Exporter.Geneva/CHANGELOG.md:18]: I would suggest that we capture the expectation from the users' perspective. Maybe something like (to be frank I haven't spent much time polishing the message):

"Changed the behavior of during initialization on Linux. Before this change, the exporter initialization would throw exception if the target Unix Domain Socket does not exist; after this change, the exporter initialization would return success and the exporting background thread will try to establish the connection."

Comment (arunmk-ms) [reply src/OpenTelemetry.Exporter.Geneva/CHANGELOG.md]: Updated change log as per your suggestion. Thank you!

File: src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/Transport/UnixDomainSocketDataTransport.cs
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Geneva/MsgPackExporter/Transport/UnixDomainSocketDataTransport.cs:46]: we can keep the same code as before, with just one change: wrap this line inside try..catch, and log the fact that UDS was unavailable at startup.

**Comment by cijothomas** [comment]
Thanks!

Given we are in RC state right now, we might hold off merging until 1.4.0 stable is out, so we'll get some time to test this.

**Comment by wsugarman** [comment]
I am concerned a bit that this PR will further reduce any "fail-fast" behavior that may save telemetry from being dropped. If my understanding of the code is accurate, today if there are any errors when sending telemetry batches to the socket, then those batches will simply be dropped. However today, at the very least, the constructor will throw exceptions and fail the application if the socket is misconfigured or not ready.

I can create a separate issue for it, but I'd love to see more resiliency in the exporter with regards to transient failures -- perhaps retries or even some preservation of unsent batches while managing a reasonable amount of backpressure.

**Comment by cijothomas** [comment]
> I am concerned a bit that this PR will further reduce any "fail-fast" behavior that may save telemetry from being dropped. If my understanding of the code is accurate, today if there are any errors when sending telemetry batches to the socket, then those batches will simply be dropped. However today, at the very least, the constructor will throw exceptions and fail the application if the socket is misconfigured or not ready.
> 
> I can create a separate issue for it, but I'd love to see more resiliency in the exporter with regards to transient failures -- perhaps retries or even some preservation of unsent batches while managing a reasonable amount of backpressure.

You concern is valid, and we did consider it before deciding to take this in.
1. In windows also, where we use ETW, the telemetry is simply lost if agent is not ready. This PR makes Linux behavior somewhat same - if agent is not available, telemetry is simply lost. 
2. This is consistent with OTLP Exporters - they dont fail at startup.

Agree about the data loss issues/retry etc - We are adding it to OTLP Exporters - https://github.com/open-telemetry/opentelemetry-dotnet/issues/4115 and could likely re-use idea from it for the Linux side.

(Long term, Linux would move away from UDS towards user_events (https://docs.kernel.org/trace/user_events.html) in Linux, once its get ready in linux distributions), so it'll be consistent with Windows (no batching etc!)

**Comment by dustinburson** [comment]
> 

Any consideration to allowing this behavior to be configurable?  Also is there any recommendation on how to monitor this in the process?   For example, the code in question has this line ```ExporterEventSource.Log.ExporterException("UDS unavailable at startup.", ex);```.  I assume it is possible to look for these events and respond?  If so is the a similar mechanism to detect if the UDS is becomes available?  I am thinking of a scenario where we might be okay with this for a limited time but if we detect failures for an extended period (say 5 minutes) we might want to take some action to avoid a process running for an extended prior of time without telemetry.

**Comment by cijothomas** [comment]
> > 
> 
> Any consideration to allowing this behavior to be configurable? Also is there any recommendation on how to monitor this in the process? For example, the code in question has this line `ExporterEventSource.Log.ExporterException("UDS unavailable at startup.", ex);`. I assume it is possible to look for these events and respond? If so is the a similar mechanism to detect if the UDS is becomes available? I am thinking of a scenario where we might be okay with this for a limited time but if we detect failures for an extended period (say 5 minutes) we might want to take some action to avoid a process running for an extended prior of time without telemetry.


_`ExporterEventSource.Log.ExporterException("UDS unavailable at startup.", ex);`. I assume it is possible to look for these events and respond?_

https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry#self-diagnostics is one option. Or any other tools like perfview, logman, or own listener. (We do not support routing this to iLogger, but thats something which might be added in the future)

_If so is the a similar mechanism to detect if the UDS is becomes available?_

The exporters keeps attempting for each Export, so if UDS is unavailable for some time, then becomes available, those telemetry from that down-window is lost, the rest will continue to flow.

_we might want to take some action to avoid a process running for an extended prior of time without telemetry._
The best option right now is to write own listener to the Exporter eventsource, and react to it (like sending some metrics saying UDS unavailable...). 

_Any consideration to allowing this behavior to be configurable?_
As mentioned in the previous comment, the goal for GenevaExporter long term is to only have native-to-OS transport. i.e ETW/TLD in windows and user_events in Linux. We are hesistant to keep adding more configs/switches which are not aligned with the long term direction.

**Comment by reyang** [comment]
> Any consideration to allowing this behavior to be configurable? Also is there any recommendation on how to monitor this in the process? For example, the code in question has this line `ExporterEventSource.Log.ExporterException("UDS unavailable at startup.", ex);`. I assume it is possible to look for these events and respond? If so is the a similar mechanism to detect if the UDS is becomes available? I am thinking of a scenario where we might be okay with this for a limited time but if we detect failures for an extended period (say 5 minutes) we might want to take some action to avoid a process running for an extended prior of time without telemetry.

I suggest us solving this problem outside the SDK/Exporter. For example, using ETW there is no easy for the SDK to know whether the data is received by the downstream or not (e.g. ETW does not tell the sender if the data is dropped due to buffer full), and such design comes with good reason.

I think something like heartbeat/QoS is a better direction.

File: src/OpenTelemetry.Exporter.Geneva/CHANGELOG.md
Comment (Kielek) [inline src/OpenTelemetry.Exporter.Geneva/CHANGELOG.md:36]: After merges with main it was moved to relesed section,. It should be under unreleased.

File: test/OpenTelemetry.Exporter.Geneva.Tests/GenevaTraceExporterTests.cs
Comment (utpilla) [inline test/OpenTelemetry.Exporter.Geneva.Tests/GenevaTraceExporterTests.cs:362]: Could you add a comment here saying "GenevaExporter would not throw if it was not able to connect to the UDS socket in ctor. It would keep attempting to connect to the socket when sending telemetry."

**Review by utpilla — APPROVED** [review]

File: test/OpenTelemetry.Exporter.Geneva.Tests/GenevaTraceExporterTests.cs
Comment (cijothomas) [inline test/OpenTelemetry.Exporter.Geneva.Tests/GenevaTraceExporterTests.cs:362]:
```suggestion

                // GenevaExporter would not throw if it was not able to connect to the UDS socket in ctor. It would
```

File: src/OpenTelemetry.Exporter.Geneva/CHANGELOG.md
Comment (cijothomas) [inline src/OpenTelemetry.Exporter.Geneva/CHANGELOG.md:11]:
```suggestion

```

**Comment by utpilla** [comment]
@arunmk-ms We released the stable 1.4.0 version of GenevaExporter. Could you please resolve the merge conflicts, and we can get this PR merged?

**Comment by arunmk-ms** [comment]
@utpilla done

**Comment by utpilla** [comment]
Thanks for your contribution @arunmk-ms !

**Pull request merged**
